src/tls_tpm.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2020 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "tls_tpm.h"
+#include "log.h"
+#include "tls.h"
+#include <openssl/engine.h>
+#include <openssl/ssl.h>
+#include <openssl/ui.h>
+#include <stdio.h>
+#include <string.h>
+
+/* Forward declarations */
+static UI_METHOD*
+enftun_tls_tpm_ssl_ui_passthrough(void);
+
+static ENGINE*
+enftun_tls_tpm_engine_init()
+{
+    ENGINE* eng                = NULL;
+    const char* engine_names[] = {"tpm2tss", "tpm2"};
+    size_t engine_index;
+
+    /* Init the engine */
+    ENGINE_load_builtin_engines();
+
+    for (engine_index = 0;
+         engine_index < sizeof(engine_names) / sizeof(engine_names[0]);
+         engine_index++)
+    {
+        eng = ENGINE_by_id(engine_names[engine_index]);
+        if (eng)
+        {
+            enftun_log_info("INFO: Using engine %s",
+                            engine_names[engine_index]);
+            break;
+        }
+    }
+    if (!eng)
+    {
+        enftun_log_error("Could not find a TSS engine\n");
+        goto out;
+    }
+
+    if (!ENGINE_init(eng))
+    {
+        enftun_log_ssl_error("Could not init engine %s",
+                             engine_names[engine_index]);
+        ENGINE_free(eng);
+        goto out;
+    }
+
+out:
+    return eng;
+}
+
+int
+enftun_tls_tpm_use_key(struct enftun_tls* tls, const char* key_file)
+{
+    ENGINE* e;
+    EVP_PKEY* key   = NULL;
+    UI_METHOD* meth = NULL;
+    int ret         = 0;
+
+    /* Having XTT enabled causes library conflicts due to xaptum-tpm */
+#ifdef USE_XTT
+    enftun_log_error("TPM key not supported when XTT is enabled.");
```

Comment:
Move this to after the `enftun_tls_tpm_is_tpm_key()` check.  No sense yelling at the user if they weren't trying a TSS key.

Reply:
Changed